src/payment/strategies/affirm/affirm-payment-strategy.ts

```diff
@@ -0,0 +1,273 @@
+import { CheckoutStore, InternalCheckoutSelectors } from '../../../checkout';
+import { MissingDataError, MissingDataErrorType, NotInitializedError, NotInitializedErrorType } from '../../../common/error/errors';
+import { OrderActionCreator, OrderRequestBody } from '../../../order';
+import { RemoteCheckoutActionCreator } from '../../../remote-checkout';
+import { PaymentArgumentInvalidError } from '../../errors';
+import PaymentActionCreator from '../../payment-action-creator';
+import { PaymentInitializeOptions, PaymentRequestOptions } from '../../payment-request-options';
+import PaymentStrategy from '../payment-strategy';
+
+import { AffirmAddress, AffirmDiscount, AffirmItem, AffirmRequestData, SCRIPTS_DEFAULT } from './affirm';
+import affirmJs from './affirmJs';
+declare let affirm: any;
+
+export default class AffirmPaymentStrategy implements PaymentStrategy {
+
+    constructor(
+        private _store: CheckoutStore,
+        private _orderActionCreator: OrderActionCreator,
+        private _paymentActionCreator: PaymentActionCreator,
+        private _remoteCheckoutActionCreator: RemoteCheckoutActionCreator
+    ) { }
+
+    initialize(options: PaymentInitializeOptions): Promise<InternalCheckoutSelectors> {
+        const state = this._store.getState();
+        const paymentMethod = state.paymentMethods.getPaymentMethod(options.methodId, options.gatewayId);
+
+        if (!paymentMethod ) {
+            throw new MissingDataError(MissingDataErrorType.MissingPaymentMethod);
+        }
+
+        const { testMode } = paymentMethod.config;
+        const { publicKey } = paymentMethod.initializationData;
+
+        affirmJs(publicKey, this._getScriptURI(testMode));
+
+        return Promise.resolve(this._store.getState());
+    }
+
+    execute(payload: OrderRequestBody, options?: PaymentRequestOptions): Promise<InternalCheckoutSelectors> {
+        const paymentId = payload.payment && payload.payment.methodId;
+        const { useStoreCredit } = payload;
+        const state = this._store.getState();
+        const config = state.config.getStoreConfig();
+
+        if (!paymentId) {
+            throw new PaymentArgumentInvalidError(['payment.methodId']);
+        }
+
+        if (!config) {
+            throw new MissingDataError(MissingDataErrorType.MissingCheckoutConfig);
+        }
+
+        return this._store.dispatch(this._remoteCheckoutActionCreator.initializePayment(paymentId, { useStoreCredit }))
+            .then(affirm.checkout(this._initializeCheckout(useStoreCredit)))
+            .then(affirm.checkout.open())
+            .then(affirm.ui.ready(
+                () => {
+                    affirm.ui.error.on('close', () => {
+                        window.location.href = `${config.links.checkoutLink}.php?action=set_external_checkout&provider=affirm&status=cancelled`;
```

Comment:
Also, I suppose Affirm presents some custom UI form to the shopper and asks them to interact with it before redirecting them elsewhere? If so, and if there could be a point of failure after the `initialize` promise is resolved, we should provide an error callback so there's a way for us to inform the client / app about the failure so it can decide how to handle it.